### PR TITLE
Move NER output formatting to server 

### PR DESCRIPTION
**Pull request by magdyksaleh**

### CODE REVIEW

File: router/client/src/sharded_client.rs [hunk 1]
```diff
@@ -268,3 +307,68 @@ fn merge_generations(
...
+        re.find_iter(&tokenizer.decode(input_ids, true).unwrap())
+            .map(|m| m.as_str().to_string())
+            .collect()
+    };
+
+    let mut ner_results = Vec::new();
+    let mut current_entity: Option<Entity> = None;
+    for (i, ((token, token_class), score)) in tokens
+        .iter()
+        .zip(predicted_token_class.iter())
```
Comment (jeffreyftang) [inline router/client/src/sharded_client.rs:332]: Btw, looks like there's a nice macro for zipping multiple iterators: https://stackoverflow.com/a/29669741

Comment (magdyksaleh) [reply router/client/src/sharded_client.rs:332]: oooh good find

File: router/client/src/sharded_client.rs [hunk 2]
```diff
@@ -174,8 +184,37 @@ impl ShardedClient {
             .iter_mut()
             .map(|client| Box::pin(client.classify(batch.clone())))
             .collect();
-        let results: Result<Vec<Vec<EntityList>>> = join_all(futures).await.into_iter().collect();
-        Ok(results?.into_iter().flatten().collect())
+        let results: Result<Vec<Vec<ClassifyPredictionList>>> =
+            join_all(futures).await.into_iter().collect();
+
+        let flat_results: Vec<ClassifyPredictionList> = results?.into_iter().flatten().collect();
+        let threaded_futures: Vec<_> = flat_results
```
Comment (tgaddair) [inline router/client/src/sharded_client.rs:191]: I would be very curious to time this block of code to see how long it takes. My concern with this approach is that even though we're parallelizing the `format_ner_outputs` operation, we're doing so in a way that is blocking the batching thread, so during this period of time, the GPU is entirely idle.

Instead, we should move this logic into the request thread here:

https://github.com/predibase/lorax/blob/main/router/src/infer.rs#L659-L661

In that way, we do not block the batching thread so we can move on to the next batch and keep the GPU busy.

Does that make sense? Happy to quickly sync on this if needed.

Comment (magdyksaleh) [reply router/client/src/sharded_client.rs]: addressed

File: router/src/server.rs [hunk 3]
```diff
@@ -1079,6 +1079,7 @@ pub async fn run(
     struct ApiDoc;
 
     let cloned_tokenizer = tokenizer.clone().map(|t| Arc::new(Mutex::new(t)));
+    let arc_tokenizer = tokenizer.clone().map(Arc::new);
```
Comment (tgaddair) [inline router/src/server.rs:1082]: Is it not sufficient to use the `cloned_tokenizer` here?

**Review by tgaddair — APPROVED** [review]
Nice! Do you have benchmarking results comparing this change with the previous version (decoding in the batching thread)?

**Comment by magdyksaleh** [comment]
> Nice! Do you have benchmarking results comparing this change with the previous version (decoding in the batching thread)?

Yep: slightly better performance when moving the output formatting out of the batching task. https://predibase.slack.com/archives/C03MMJ8J5GF/p1722357582611049?thread_ts=1722331652.853819&cid=C03MMJ8J5GF

**Pull request merged**